collector/bcache_linux.go

```diff
@@ -0,0 +1,307 @@
+			desc:            "Count of times readahead occurred.",
+			value:           float64(ps.CacheReadaheads),
+			metricType:      prometheus.CounterValue,
+			extraLabel:      label,
+			extraLabelValue: labelValue,
+		},
+	}
+	return metrics
+}
+
+// UpdateBcacheStats collects statistics for one bcache ID.
+func (c *bcacheCollector) updateBcacheStats(ch chan<- prometheus.Metric, s *bcache.Stats) {
+
+	const (
+		subsystem = "bcache"
+	)
+
+	var (
+		devLabel   = []string{"uuid"}
+		allMetrics []bcacheMetric
+		metrics    []bcacheMetric
+	)
+
+	allMetrics = []bcacheMetric{
+		// metrics in /sys/fs/bcache/<uuid>/
+		{
+			name:       "average_key_size_sectors",
+			desc:       "Average data per key in the btree (sectors).",
+			value:      float64(s.Bcache.AverageKeySize),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_cache_size_bytes",
+			desc:       "Amount of memory currently used by the btree cache.",
+			value:      float64(s.Bcache.BtreeCacheSize),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "cache_available_percent",
+			desc:       "Percentage of cache device without dirty data, useable for writeback (may contain clean cached data).",
+			value:      float64(s.Bcache.CacheAvailablePercent),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "congested",
+			desc:       "Congestion.",
+			value:      float64(s.Bcache.Congested),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "root_usage_percent",
+			desc:       "Percentage of the root btree node in use (tree depth increases if too high).",
+			value:      float64(s.Bcache.RootUsagePercent),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "tree_depth",
+			desc:       "Depth of the btree.",
+			value:      float64(s.Bcache.TreeDepth),
+			metricType: prometheus.GaugeValue,
+		},
+		// metrics in /sys/fs/bcache/<uuid>/internal/
+		{
+			name:       "active_journal_entries",
+			desc:       "Number of journal entries that are newer than the index.",
+			value:      float64(s.Bcache.Internal.ActiveJournalEntries),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_nodes",
+			desc:       "Total nodes in the btree.",
+			value:      float64(s.Bcache.Internal.BtreeNodes),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "btree_read_average_duration_nanoseconds",
+			desc:       "Average btree read duration.",
+			value:      float64(s.Bcache.Internal.BtreeReadAverageDurationNanoSeconds),
+			metricType: prometheus.GaugeValue,
+		},
+		{
+			name:       "cache_read_races",
+			desc:       "Counts instances where while data was being read from the cache, the bucket was reused and invalidated - i.e. where the pointer was stale after the read completed.",
+			value:      float64(s.Bcache.Internal.CacheReadRaces),
+			metricType: prometheus.CounterValue,
+		},
+	}
+
+	metrics = bcachePeriodStatsToMetric(&s.Bcache.Total, "all")
```

Comment:
In our experience, metrics become a lot easier to handle if time series with different label values can be summed up and return the total, e.g. `sum without(backing_device) (node_bcache_cache_hits_total)`. This won't longer be possible with this `all` label.

Do we really need the totals as extra metric? Wouldn't it be enough to calculate the total with Prometheus side? That's in general our guideline: don't export sums, only export the individual values and calculate sums on the server.

If these Bcache.Total metrics can't be calculated, it might make sense to use a different metric name.